# **Market Basket Analysis**

Market basket analysis or MBA is a data mining technique that is used to determine the purchase habits of the consumers. The aim of the technique is to uncover the set of products that are frequently brought together. 

Association rules are 'if-then' statments, that reflects the probability of relationships among dataset. It is used in Market basket analysis to determine the relationship between products. This asssociation between products is used by retailers in various marketing decisions. Association rule helps in developing following strategies:
- Product placement
- Personalised push notifications
- Catalog design
- Cross-selling & Up-selling

Strategic and mindful placement of products in the aisle not only save consumer's time but it also encourages consumers to buy related products. For example, consumer buying cereal is more likely to buy milk if milk aisle is closer to the cereal section. 

The 'if' component of the association rule is known as 'antecedent' and the 'then' component is 'consequent'. These components are disjoint, they just reflect co-occurence and not casuality. 

The strength of association rule is measured by support, confidence and lift ratio.

* Support: Support measures the fraction of product's occurence. In simple terms, it reflects the popularity of the product. The number of times a product occurs in transactions. For example, `milk` will occur in more number of transactions then  `shaving cream`. Hence `milk` generally have higher support than `shaving cream`.

* Confidence: Confidence is the ratio of the number of transactions that contain both antecedent & consequent to the number of transactions that contain all antecedent itemsets. As the name suggest, confidence reflects the reliability of the rule and conditional probability that the consequent will occur given the occurrence of the antecedent.

* Lift: Lift is the ratio of confidence of the rule to the expected confidence. In simple terms, it measures the importance and quality of the rule.

In [1]:
import pandas as pd

from mlxtend.frequent_patterns import apriori
from mlxtend.frequent_patterns import association_rules

pd.set_option('display.max_colwidth', None)

In [2]:
order_prior = pd.read_csv('order_products__prior.csv')
products = pd.read_csv('products.csv')

In [ ]:
order_prior.head()

order_id  product_id  add_to_cart_order  reordered
0         2       33120                  1          1
1         2       28985                  2          1
2         2        9327                  3          0
3         2       45918                  4          1
4         2       30035                  5          0

In [3]:
order_prior.product_id.nunique()

49677

We have 49,677 products in our dataset. Apriori algorithm looks at the database multiple times to determine the frequency of the itemsets. So, it can be slow and inefficent on large datsets. Hence, we look at only 200 most frequently bought products. 

In [4]:
#calculating top 200 most bought products

top_200_products = order_prior.groupby('product_id')['order_id'].count().sort_values(ascending = False)[:200].reset_index()

In [5]:
#creating a dataframe of only top 200 products

df = order_prior[order_prior.product_id.isin(top_200_products.product_id.to_list())]

In [6]:
#merging df with products data on product_id to get product name
df = df.merge(products, on = 'product_id', how = 'left')

#keeping only the relevant columns
df= df[['order_id','product_name','reordered']].set_index('order_id') 

In [ ]:
df.head()

product_name  reordered
order_id                                                                  
2                                            Organic Egg Whites          1
2                                         Michigan Organic Kale          1
2                                                       Carrots          1
3         Total 2% with Strawberry Lowfat Greek Strained Yogurt          1
3                                        Unsweetened Almondmilk          1

In [7]:
x = df.pivot_table(columns='product_name', values='reordered', index='order_id').reset_index().fillna(0).set_index('order_id')

#converting data type from float to int
x = x.astype('int32')

In [ ]:
x

product_name  100% Raw Coconut Water  100% Recycled Paper Towels  \
order_id                                                           
2                                  0                           0   
3                                  0                           0   
4                                  0                           0   
5                                  0                           0   
7                                  0                           0   
...                              ...                         ...   
3421078                            0                           0   
3421080                            0                           0   
3421081                            0                           0   
3421082                            0                           0   
3421083                            0                           0   

product_name  100% Whole Wheat Bread  2% Reduced Fat Milk  \
order_id                                                    
2                                  0                    0   
3                                  0                    0   
4                                  0                    0   
5                                  0                    1   
7                                  0                    0   
...                              ...                  ...   
3421078                            0                    0   
3421080                            0                    0   
3421081                            0                    0   
3421082                            0                    0   
3421083                            0                    0   

product_name  Apple Honeycrisp Organic  Asparagus  Baby Spinach  \
order_id                                                          
2                                    0          0             0   
3                                    0          0             0   
4                                    0          0             0   
5                                    0          0             0   
7                                    0          0             0   
...                                ...        ...           ...   
3421078                              0          0             0   
3421080                              0          0             0   
3421081                              0          0             0   
3421082                              0          0             0   
3421083                              0          0             0   

product_name  Bag of Organic Bananas  Banana  Bartlett Pears  ...  \
order_id                                                      ...   
2                                  0       0               0  ...   
3                                  0       0               0  ...   
4                                  0       0               0  ...   
5                                  1       0               0  ...   
7                                  0       0               0  ...   
...                              ...     ...             ...  ...   
3421078                            0       1               0  ...   
3421080                            0       0               0  ...   
3421081                            0       0               0  ...   
3421082                            0       0               0  ...   
3421083                            0       1               0  ...   

product_name  Unsweetened Almondmilk  \
order_id                               
2                                  0   
3                                  1   
4                                  0   
5                                  0   
7                                  0   
...                              ...   
3421078                            0   
3421080                            0   
3421081                            0   
3421082                            0   
3421083                            0   

product_name  Unsweetened Orig

In [ ]:
frequent_sets = apriori(x, min_support=0.003, use_colnames=True )

In [ ]:
frequent_sets

support  \
0    0.011726   
1    0.006451   
2    0.017559   
3    0.011275   
4    0.024482   
5    0.016287   
6    0.005458   
7    0.123727   
8    0.156115   
9    0.009014   
10   0.005610   
11   0.012765   
12   0.005874   
13   0.013036   
14   0.010786   
15   0.004692   
16   0.010135   
17   0.005946   
18   0.018240   
19   0.005571   
20   0.009176   
21   0.009406   
22   0.004770   
23   0.004114   
24   0.026363   
25   0.006252   
26   0.009382   
27   0.008835   
28   0.007077   
29   0.018004   
30   0.004576   
31   0.007276   
32   0.005890   
33   0.008424   
34   0.009218   
35   0.010170   
36   0.010823   
37   0.006635   
38   0.013260   
39   0.005400   
40   0.006728   
41   0.020651   
42   0.014634   
43   0.004095   
44   0.006008   
45   0.022644   
46   0.009859   
47   0.009690   
48   0.012267   
49   0.004709   
50   0.041615   
51   0.005159   
52   0.006134   
53   0.013527   
54   0.037508   
55   0.004904   
56   0.018345   
57   0.006758   
58   0.005797   
59   0.006347   
60   0.009240   
61   0.007037   
62   0.052498   
63   0.019503   
64   0.008084   
65   0.019767   
66   0.004542   
67   0.073193   
68   0.010454   
69   0.005353   
70   0.005407   
71   0.008754   
72   0.015782   
73   0.024643   
74   0.005721   
75   0.006450   
76   0.010831   
77   0.008393   
78   0.003633   
79   0.008132   
80   0.004895   
81   0.016881   
82   0.005673   
83   0.021952   
84   0.013230   
85   0.004100   
86   0.005383   
87   0.007569   
88   0.008663   
89   0.004462   
90   0.024992   
91   0.020916   
92   0.006557   
93   0.029242   
94   0.013476   
95   0.011690   
96   0.011221   
97   0.016279   
98   0.021631   
99   0.004337   
100  0.005280   
101  0.023371   
102  0.066632   
103  0.005763   
104  0.006315   
105  0.014546   
106  0.004489   
107  0.014007   
108  0.011724   
109  0.009763   
110  0.022403   
111  0.007736   
112  0.004005   
113  0.023709   
114  0.008896   
115  0.004229   
116  0.010705   
117  0.004262   
118  0.013675   
119  0.041283   
120  0.006219   
121  0.014870   
122  0.016187   
123  0.006555   
124  0.015096   
125  0.011882   
126  0.007399   
127  0.007359   
128  0.005711   
129  0.008439   
130  0.007807   
131  0.004896   
132  0.016830   
133  0.005816   
134  0.008331   
135  0.005045   
136  0.009804   
137  0.080619   
138  0.003173   
139  0.016552   
140  0.004746   
141  0.017184   
142  0.005794   
143  0.044848   
144  0.007727   
145  0.017875   
146  0.004481   
147  0.030969   
148  0.005009   
149  0.028263   
150  0.020244   
151  0.006189   
152  0.005010   
153  0.005643   
154  0.005601   
155  0.008727   
156  0.009189   
157  0.015473   
158  0.009246   
159  0.013713   
160  0.006603   
161  0.013634   
162  0.010043   
163  0.006081   
164  0.006465   
165  0.021610   
166  0.007262   
167  0.006237   
168  0.004450   
169  0.009153   
170  0.012895   
171  0.010884   
172  0.004488   
173  0.009557   
174  0.012337   
175  0.022917   
176  0.017843   
177  0.039087   
178  0.004049   
179  0.007078   
180  0.007178   
181  0.006582   
182  0.006107   
183  0.005883   
184  0.008780   
185  0.007316   
186  0.010720   
187  0.005021   
188  0.007766   
189  0.005151   
190  0.014625   
191  0.009670   
192  0.007257   
193  0.008829   
194  0.007115   
195  0.008430   
196  0.010868   
197  0.005396   
198  0.016946   
199  0.005363   
200  0.003205   
201  0.004211   
202  0.006888   
203  0.004818   
204  0.003150   
205  0.004436   
206  0.004354   
207  0.003597   
208  0.003144   
209  0.003707   
210  0.003091   
211  0.004965   
212  0.005465   
213  0.003109   
214  0.006165   
215  0.003021   
216  0.003999   
217  0.014769   
218  0.003016   
219  0.005359   
220  0.003457   
221  0.006056   
222  0.004193   
223  0.005538   
224  0.005640   
225  0.003398   
226  0.004045   
227  0.004342   
228  0.004090   
229  0.019559   
230  0.004081   
231  0.007055   
232  0.006175   
233  0.003774   
234  0.0123

We will derive the association rule on the basis of lift. Lift is more important than confidence in the association rule because it reflects the relative strength of association between 2 products. It is the ratio of the change in the probability of presence of item 1 with the knowledge that item 2 is present over the probability of presence of item 1 without the knowledge of item 2's presence (Nandakumar,2020).

For example: 
Probability of milk in the cart **with the knowledge** of presence of toothbrush is = 10/(10+4) = .7

If the probability of milk in the cart **without the knowledge** of toothbrush is 80/100 = .8

We can see above that the knowledge of toothbrush reduces the probability of milk from 0.8 to 07. Hence the lift here will be 0.7/.8 = .87 and lift less than 1 shows the less association between products.

We will set min_threshold to be 1 because lift more than 1 means that products are likly to be bought together where lift of 1 means that products have no association between tham and lift of less than 1 means that products are not likly to be bought together

In [ ]:
rules = association_rules(frequent_sets, metric='lift', min_threshold=1)
rules.sort_values(by=['lift'], ascending=False)

antecedents  \
197                    (Sparkling Water Grapefruit)   
196                          (Lime Sparkling Water)   
275                          (Organic Yellow Onion)   
274                                (Organic Garlic)   
202                                         (Limes)   
203                              (Organic Cilantro)   
319                                  (Strawberries)   
318                                   (Raspberries)   
348                           (Organic Raspberries)   
345  (Bag of Organic Bananas, Organic Strawberries)   
185                                   (Large Lemon)   
184                                         (Limes)   
333  (Bag of Organic Bananas, Organic Hass Avocado)   
336                           (Organic Raspberries)   
332   (Bag of Organic Bananas, Organic Raspberries)   
337                          (Organic Hass Avocado)   
282                                 (Organic Lemon)   
283                          (Organic Hass Avocado)   
335                        (Bag of Organic Bananas)   
334     (Organic Raspberries, Organic Hass Avocado)   
325                          (Organic Hass Avocado)   
320  (Organic Baby Spinach, Bag of Organic Bananas)   
343                          (Organic Hass Avocado)   
338  (Bag of Organic Bananas, Organic Strawberries)   
349                          (Organic Strawberries)   
344   (Bag of Organic Bananas, Organic Raspberries)   
205                                (Organic Garlic)   
204                                         (Limes)   
262                              (Organic Cucumber)   
263                          (Organic Hass Avocado)   
177                                   (Large Lemon)   
176                                (Cucumber Kirby)   
258                           (Organic Raspberries)   
259                           (Organic Blueberries)   
288                        (Organic Tomato Cluster)   
289                          (Organic Hass Avocado)   
190                                (Organic Garlic)   
191                                   (Large Lemon)   
340    (Organic Strawberries, Organic Hass Avocado)   
341                        (Bag of Organic Bananas)   
187                               (Organic Avocado)   
186                                   (Large Lemon)   
303                          (Organic Strawberries)   
302                           (Organic Raspberries)   
321    (Organic Baby Spinach, Organic Hass Avocado)   
324                        (Bag of Organic Bananas)   
296                          (Organic Strawberries)   
297                                  (Organic Kiwi)   
63                           (Organic Navel Orange)   
62                         (Bag of Organic Bananas)   
339  (Bag of Organic Bananas, Organic Hass Avocado)   
342                          (Organic Strawberries)   
346     (Organic Raspberries, Organic Strawberries)   
347                        (Bag of Organic Bananas)   
178                                (Cucumber Kirby)   
179                               (Organic Avocado)   
284                           (Organic Raspberries)   
285                          (Organic Hass Avocado)   
260                          (Organic Strawberries)   
261                           (Organic Blueberries)   
11                           (Organic Hass Avocado)   
10                       (Apple Honeycrisp Organic)   
199                               (Organic Avocado)   
198                                         (Limes)   
213                          (Organic Baby Spinach)   
212                         (Michigan Organic Kale)   
246                          (Organic Baby Spinach)   
247                    (Organic Small Bunch Celery)   
309                          (Organic Strawberries)   
308                   (Organic Whole String Cheese)   
281                          (Organic Hass Avocado)   
280          (Organic Large Extra Fancy Fuji Apple)   
210                      

In [ ]:
#adding length column to determine the number of items in the set
rules['length'] = rules['antecedents'].apply(lambda x: len(x))

In [ ]:
rules.sort_values(by=['lift', 'confidence'], ascending=[False, False])

antecedents  \
196                          (Lime Sparkling Water)   
197                    (Sparkling Water Grapefruit)   
274                                (Organic Garlic)   
275                          (Organic Yellow Onion)   
202                                         (Limes)   
203                              (Organic Cilantro)   
319                                  (Strawberries)   
318                                   (Raspberries)   
345  (Bag of Organic Bananas, Organic Strawberries)   
348                           (Organic Raspberries)   
184                                         (Limes)   
185                                   (Large Lemon)   
333  (Bag of Organic Bananas, Organic Hass Avocado)   
336                           (Organic Raspberries)   
332   (Bag of Organic Bananas, Organic Raspberries)   
337                          (Organic Hass Avocado)   
282                                 (Organic Lemon)   
283                          (Organic Hass Avocado)   
335                        (Bag of Organic Bananas)   
334     (Organic Raspberries, Organic Hass Avocado)   
325                          (Organic Hass Avocado)   
320  (Organic Baby Spinach, Bag of Organic Bananas)   
343                          (Organic Hass Avocado)   
338  (Bag of Organic Bananas, Organic Strawberries)   
344   (Bag of Organic Bananas, Organic Raspberries)   
349                          (Organic Strawberries)   
205                                (Organic Garlic)   
204                                         (Limes)   
262                              (Organic Cucumber)   
263                          (Organic Hass Avocado)   
176                                (Cucumber Kirby)   
177                                   (Large Lemon)   
259                           (Organic Blueberries)   
258                           (Organic Raspberries)   
288                        (Organic Tomato Cluster)   
289                          (Organic Hass Avocado)   
190                                (Organic Garlic)   
191                                   (Large Lemon)   
340    (Organic Strawberries, Organic Hass Avocado)   
341                        (Bag of Organic Bananas)   
186                                   (Large Lemon)   
187                               (Organic Avocado)   
303                          (Organic Strawberries)   
302                           (Organic Raspberries)   
321    (Organic Baby Spinach, Organic Hass Avocado)   
324                        (Bag of Organic Bananas)   
297                                  (Organic Kiwi)   
296                          (Organic Strawberries)   
63                           (Organic Navel Orange)   
62                         (Bag of Organic Bananas)   
339  (Bag of Organic Bananas, Organic Hass Avocado)   
342                          (Organic Strawberries)   
346     (Organic Raspberries, Organic Strawberries)   
347                        (Bag of Organic Bananas)   
178                                (Cucumber Kirby)   
179                               (Organic Avocado)   
284                           (Organic Raspberries)   
285                          (Organic Hass Avocado)   
261                           (Organic Blueberries)   
260                          (Organic Strawberries)   
11                           (Organic Hass Avocado)   
10                       (Apple Honeycrisp Organic)   
199                               (Organic Avocado)   
198                                         (Limes)   
212                         (Michigan Organic Kale)   
213                          (Organic Baby Spinach)   
247                    (Organic Small Bunch Celery)   
246                          (Organic Baby Spinach)   
308                   (Organic Whole String Cheese)   
309                          (Organic Strawberries)   
280          (Organic Large Extra Fancy Fuji Apple)   
281                          (Organic Hass Avocado)   
210                      

As we can see above there are several rules from same item sets. In these rules, support will remain same however the denominator calculation of confidence will differ. For example:

Rule 1 : ratio of transactions containing 'milk' also having 'bread, cereal, butter'

Rule 2 : ratio of transactions containing 'bread, cereal, butter' also having 'milk'

Rule 1 < Rule 2 because the number of transactions containing 'milk' with 'bread, cereal, butter' will be far high then transactions that contain all three 'bread, cereal, butter' and also 'milk'. 

We will look for this method to prune the rules.

In [ ]:
def threshold(support = .001,confidence=.005,lift=1.1, length = 1):

  ''' This function will take the given threshold or the default minimum threshold and generate the desired result'''

  return rules[ (rules['support'] >= support) &
       (rules['confidence'] >= confidence) &
       (rules['lift'] >= lift) &
       (rules['length'] == length)].sort_values(by = 'lift', ascending = False) 

In [ ]:
threshold(length = 2, confidence =.2).sort_values(by=['confidence','lift'], ascending=[False, False])

antecedents               consequents  \
334     (Organic Raspberries, Organic Hass Avocado)  (Bag of Organic Bananas)   
340    (Organic Strawberries, Organic Hass Avocado)  (Bag of Organic Bananas)   
321    (Organic Baby Spinach, Organic Hass Avocado)  (Bag of Organic Bananas)   
346     (Organic Raspberries, Organic Strawberries)  (Bag of Organic Bananas)   
327    (Organic Baby Spinach, Organic Strawberries)  (Bag of Organic Bananas)   
332   (Bag of Organic Bananas, Organic Raspberries)    (Organic Hass Avocado)   
344   (Bag of Organic Bananas, Organic Raspberries)    (Organic Strawberries)   
320  (Organic Baby Spinach, Bag of Organic Bananas)    (Organic Hass Avocado)   
338  (Bag of Organic Bananas, Organic Strawberries)    (Organic Hass Avocado)   
339  (Bag of Organic Bananas, Organic Hass Avocado)    (Organic Strawberries)   
326  (Organic Baby Spinach, Bag of Organic Bananas)    (Organic Strawberries)   

     antecedent support  consequent support   support  confidence      lift  \
334            0.007615            0.123727  0.003447    0.452633  3.658309   
340            0.011934            0.123727  0.004470    0.374553  3.027242   
321            0.009896            0.123727  0.003565    0.360258  2.911707   
346            0.009827            0.123727  0.003436    0.349607  2.825626   
327            0.010427            0.123727  0.003056    0.293130  2.369159   
332            0.012306            0.066632  0.003447    0.280099  4.203684   
344            0.012306            0.080619  0.003436    0.279176  3.462899   
320            0.014769            0.066632  0.003565    0.241395  3.622814   
338            0.018520            0.066632  0.004470    0.241361  3.622307   
339            0.019559            0.080619  0.004470    0.228530  2.834678   
326            0.014769            0.080619  0.003056    0.206948  2.566977   

     leverage  conviction  length  
334  0.002505    1.600887       2  
340  0.002993    1.401034       2  
321  0.002341    1.369728       2  
346  0.002220    1.347298       2  
327  0.001766    1.239652       2  
332  0.002627    1.296523       2  
344  0.002443    1.275459       2  
320  0.002581    1.230374       2  
338  0.003236    1.230319       2  
339  0.002893    1.191725       2  
326  0.001866    1.159294       2

Rules with 2 itemsets and minimum confidence of 20% are:
* {Organic Raspberries, Organic Hass Avocado}  --▶ {Bag of Organic Bananas}
* {Organic Strawberries, Organic Hass Avocado}  --▶ {Bag of Organic Bananas}
* {Organic Baby Spinach, Organic Hass Avocado}  --▶ {Bag of Organic Bananas}
* {Organic Raspberries, Organic Strawberries}  --▶ {Bag of Organic Bananas}
* {Organic Baby Spinach, Organic Strawberries}  --▶ {Bag of Organic Bananas}



In [ ]:
threshold(length = 1,  confidence = .30).sort_values(by=['confidence','lift'], ascending=[False, False])

antecedents               consequents  \
95                         (Bartlett Pears)                  (Banana)   
133                    (Organic Fuji Apple)                  (Banana)   
113                      (Honeycrisp Apple)                  (Banana)   
63                   (Organic Navel Orange)  (Bag of Organic Bananas)   
109                   (Granny Smith Apples)                  (Banana)   
101                        (Broccoli Crown)                  (Banana)   
105                        (Cucumber Kirby)                  (Banana)   
41                  (Organic D'Anjou Pears)  (Bag of Organic Bananas)   
59   (Organic Large Extra Fancy Fuji Apple)  (Bag of Organic Bananas)   
167                   (Seedless Red Grapes)                  (Banana)   

     antecedent support  consequent support   support  confidence      lift  \
95             0.009014            0.156115  0.003630    0.402737  2.579742   
133            0.024992            0.156115  0.009735    0.389525  2.495112   
113            0.022644            0.156115  0.008058    0.355858  2.279456   
63             0.010705            0.123727  0.003774    0.352590  2.849735   
109            0.009218            0.156115  0.003221    0.349365  2.237863   
101            0.010786            0.156115  0.003671    0.340378  2.180296   
105            0.026363            0.156115  0.008748    0.331808  2.125404   
41             0.013230            0.123727  0.004193    0.316903  2.561303   
59             0.022403            0.123727  0.007055    0.314907  2.545168   
167            0.021610            0.156115  0.006603    0.305562  1.957284   

     leverage  conviction  length  
95   0.002223    1.412921       1  
133  0.005833    1.382342       1  
113  0.004523    1.310091       1  
63   0.002450    1.353506       1  
109  0.001781    1.297017       1  
101  0.001987    1.279345       1  
105  0.004632    1.262938       1  
41   0.002556    1.282795       1  
59   0.004283    1.279057       1  
167  0.003230    1.215205       1

Bananas and Bag of Organic Bananas is natural consequent for fruits and vegetables. Let's see the association between products excluding these two products from the set.

In [8]:
y = x.drop(['Banana', 'Bag of Organic Bananas'], axis=1)

In [9]:
y.head()

product_name  100% Raw Coconut Water  100% Recycled Paper Towels  \
order_id                                                           
2                                  0                           0   
3                                  0                           0   
4                                  0                           0   
5                                  0                           0   
7                                  0                           0   

product_name  100% Whole Wheat Bread  2% Reduced Fat Milk  \
order_id                                                    
2                                  0                    0   
3                                  0                    0   
4                                  0                    0   
5                                  0                    1   
7                                  0                    0   

product_name  Apple Honeycrisp Organic  Asparagus  Baby Spinach  \
order_id                                                          
2                                    0          0             0   
3                                    0          0             0   
4                                    0          0             0   
5                                    0          0             0   
7                                    0          0             0   

product_name  Bartlett Pears  Blackberries  Blueberries  ...  \
order_id                                                 ...   
2                          0             0            0  ...   
3                          0             0            0  ...   
4                          0             0            0  ...   
5                          0             0            0  ...   
7                          0             0            0  ...   

product_name  Unsweetened Almondmilk  \
order_id                               
2                                  0   
3                                  1   
4                                  0   
5                                  0   
7                                  0   

product_name  Unsweetened Original Almond Breeze Almond Milk  \
order_id                                                       
2                                                          0   
3                                                          0   
4                                                          0   
5                                                          0   
7                                                          0   

product_name  Unsweetened Vanilla Almond Milk  \
order_id                                        
2                                           0   
3                                           0   
4                                           0   
5                                           0   
7                                           0   

product_name  Vanilla Almond Breeze Almond Milk  Watermelon Chunks  \
order_id                                                             
2                                             0                  0   
3                                             0                  0   
4                                             0                  0   
5                                             0                  0   
7                                             0                  0   

product_name  Whipped Cream Cheese  Whole Milk  Yellow Bell Pepper  \
order_id                                                             
2                                0           0                   0   
3                                0           0                   0   
4                                0           0                   0   
5                                0           0                   0   
7                                0           0                   0   

product_name  Yellow Onions  \
order_id                      
2                         0   
3                         0

In [ ]:
frequent_set_2 = apriori(y, min_support=0.003, use_colnames=True )

In [13]:
rules_2 = association_rules(frequent_set_2, metric='lift', min_threshold=1)

In [24]:
rules_2.sort_values(by=['confidence','lift'], ascending=[False, False]).head(20)

antecedents                   consequents  \
27                   (Lime Sparkling Water)  (Sparkling Water Grapefruit)   
112                         (Organic Lemon)        (Organic Hass Avocado)   
133                   (Organic Raspberries)        (Organic Strawberries)   
126                          (Organic Kiwi)        (Organic Strawberries)   
90                    (Organic Blueberries)        (Organic Strawberries)   
32                       (Organic Cilantro)                       (Limes)   
92                       (Organic Cucumber)        (Organic Hass Avocado)   
138           (Organic Whole String Cheese)        (Organic Strawberries)   
148                           (Raspberries)                (Strawberries)   
118                (Organic Tomato Cluster)        (Organic Hass Avocado)   
134               (Organic Red Bell Pepper)        (Organic Strawberries)   
94                       (Organic Cucumber)        (Organic Strawberries)   
43                  (Michigan Organic Kale)        (Organic Baby Spinach)   
77             (Organic Small Bunch Celery)        (Organic Baby Spinach)   
128  (Organic Large Extra Fancy Fuji Apple)        (Organic Strawberries)   
62                       (Organic Cucumber)        (Organic Baby Spinach)   
105                        (Organic Garlic)        (Organic Yellow Onion)   
115                   (Organic Raspberries)        (Organic Hass Avocado)   
84                       (Organic Zucchini)        (Organic Baby Spinach)   
3                (Apple Honeycrisp Organic)        (Organic Hass Avocado)   

     antecedent support  consequent support   support  confidence       lift  \
27             0.013527            0.022917  0.003886    0.287278  12.535775   
112            0.023709            0.066632  0.005860    0.247175   3.709565   
133            0.041283            0.080619  0.009827    0.238035   2.952579   
126            0.014007            0.080619  0.003253    0.232210   2.880332   
90             0.024643            0.080619  0.005464    0.221719   2.750198   
32             0.016881            0.037508  0.003669    0.217363   5.795179   
92             0.021952            0.066632  0.004737    0.215803   3.238743   
138            0.017875            0.080619  0.003786    0.211836   2.627615   
148            0.015473            0.039087  0.003199    0.206748   5.289370   
118            0.016552            0.066632  0.003364    0.203237   3.050148   
134            0.014870            0.080619  0.003020    0.203092   2.519151   
94             0.021952            0.080619  0.004312    0.196428   2.436493   
43             0.018345            0.073193  0.003589    0.195662   2.673226   
77             0.016830            0.073193  0.003262    0.193852   2.648494   
128            0.022403            0.080619  0.004300    0.191920   2.380569   
62             0.021952            0.073193  0.004127    0.187990   2.568402   
105            0.029242            0.030969  0.005464    0.186840   6.033198   
115            0.041283            0.066632  0.007615    0.184462   2.768382   
84             0.028263            0.073193  0.005189    0.183607   2.508526   
3              0.024482            0.066632  0.004436    0.181203   2.719465   

     leverage  conviction  length  
27   0.003576    1.370918       1  
112  0.004280    1.239821       1  
133  0.006499    1.206591       1  
126  0.002123    1.197438       1  
90   0.003477    1.181297       1  
32   0.003036    1.229807       1  
92   0.003275    1.190222       1  
138  0.002345    1.166485       1  
148  0.002594    1.211359       1  
118  0.002261    1.171450       1  
134  0.001821    1.153685       1  
94   0.002542    1.144118       1  
43   0.002247    1.152260       1  
77   0.002031    1.149673       1  
128  0.002493    1.137734       1  
62   0.002520    1.141373       1  
105  0.004558    1.191685       1  
115  0.004864    1.144482       1  
84   0.003121    1.135246       1  
3    0.002805

**Observation:**

* Without bananas, length of itemsets does not exceed 1.

* Interestingly, organic and non-organic products seldomly occur together.